src/libopensc/pkcs15-eoi.c

```diff
@@ -0,0 +1,148 @@
+			/* Read the file that contains serial and encrypted CAN */
+			if (sc_pkcs15_read_file(p15card, &auth_info->path, &privdata->enc_can.value, &privdata->enc_can.len, 0) == SC_SUCCESS) {
+				/* File should be 24 bytes long */
+				if (privdata->enc_can.len != 24)
+					LOG_FUNC_RETURN(card->ctx, SC_ERROR_CORRUPTED_DATA);
+				if (strlen(p15card->tokeninfo->serial_number) != 20)
+					LOG_FUNC_RETURN(card->ctx, SC_ERROR_CORRUPTED_DATA);
+				/* First 8 bytes are used as serial number */
+				sc_bin_to_hex(privdata->enc_can.value, 8, &p15card->tokeninfo->serial_number[4], 17, 0);
+			}
+			/* Do not add "Card CAN" to the list of PIN paths to ignore, otherwise the 2nd PKCS#15 app can not access it */
+			auth_info = NULL;
+			/* Mark "Card CAN" as NOT a PIN object, so that it doesn't get it's own PKCS#11 slot */
+			objs[i]->type &= ~SC_PKCS15_TYPE_AUTH_PIN;
+		}
+		/*
+		 * For some reason QES app has "Norm PUK" not flagged as unblocking PIN and thus "Norm PUK" appears as a slot in
+		 * PKCS#11. Flag it as unblockingPin, so it doesn't appear as a separate slot.
+		 */
+		if (auth_info && auth_info->auth_id.len == 8 && !strcmp((char*)auth_info->auth_id.value, "Norm PUK")) {
+			auth_info->attrs.pin.flags |= SC_PKCS15_PIN_FLAG_UNBLOCKING_PIN;
+		}
+		if (auth_info) {
+			privdata->pin_paths[j++] = &auth_info->path;
+		}
+	}
+
+	/*
+	 * Private key objects:
+	 * 1) Rename "Card PIN" to "Norm PIN" as it's the later name that is used throughout the PKCS#15 objects
+	 * 2) Add the key references to the prkey_mappings array, as it seems that eOI expects them counted from 0xA0 up (starting from 1 within each app)
+	 *    Currently there are 3 private keys on the card
+	 *     key_ref
+	 *       2 - for pinless entry (Prijava brez PIN-a), maps to 0xA1
+	 *       1 - for authentication in QES app (Podpis in prijava), maps to 0xA1
+	 *       3 - for signing in QES app (Podpis in prijava), maps to 0xA2
+	 */
+	memset(&sk, 0, sizeof(sk));
+	sk.class_mask = SC_PKCS15_SEARCH_CLASS_PRKEY;
+	len = sc_pkcs15_search_objects(p15card, &sk, (struct sc_pkcs15_object **)&objs, MAX_OBJECTS);
+	/*
+	 * If both PKCS#15 apps are enabled, prkey_mappings can already be partially filled up from the first PKCS#15 app
+	 * as the privdata is shared between both apps which use the same driver
+	  */
+	for (j = 0; privdata->prkey_mappings[j][1] != 0; j++) {
+		/* NOP */
+	}
+	for (i = 0; i < len; i++) {
+		struct sc_pkcs15_prkey_info *prkey_info = (struct sc_pkcs15_prkey_info *)objs[i]->data;
+		if ((objs[i]->auth_id.len == 8) && !strncmp((char*)objs[i]->auth_id.value, "Card PIN", 8)) {
+			memcpy(objs[i]->auth_id.value, "Norm PIN", 8);
+		}
+		if (prkey_info) {
+			privdata->prkey_mappings[j][0] = prkey_info->key_reference;
+			privdata->prkey_mappings[j++][1] = 0xA0 + (i + 1);
+		}
+	}
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+#else
+
+#include "pkcs15.h"
+
+int sc_pkcs15emu_eoi_init_ex(struct sc_pkcs15_card *p15card, struct sc_aid *aid)
+{
+	return SC_ERROR_WRONG_CARD;
```

Comment:
please use LOG_FUNC_RETURN here so that we can easily identify that the dependencies are missing in case of an error.

Reply:
Should an acompanying LOG_FUNC_CALLED also be added, or is just LOG_FUNC_RETURN enough?

Reply:
We're not so strict, but for debugging it's helpful to see what's going on. I would add only one log entry, but do as you like